### PR TITLE
Support custom culture in RAR

### LINKED ISSUE #10536: [Bug]: Satellite assemblies with un-predefined cultures are not copied correctly into depended projects
### Issue Description

Satellite assemblies with un-predefined cultures in a project are not copied correctly into the projects that depends on it.

### Steps to Reproduce

Assume that `yue` is not a predefined language in the system.

1. Create a project (Project A) with two resources named `Test.resx` and `Test.yue.resx` respectively.
2. Add the following metadata into Project A:
   ```xml
   <PropertyGroup>
   	<RespectAlreadyAssignedItemCulture>True</RespectAlreadyAssignedItemCulture>
   </PropertyGroup>
   <ItemGroup>
   	<EmbeddedResource Update="Test.resx">
   		<Generator>ResXFileCodeGenerator</Generator>
   	</EmbeddedResource>
   	<EmbeddedResource Update="Test.yue.resx">
   		<Culture>yue</Culture>
   		<LogicalName>Test.yue.resources</LogicalName>
   	</EmbeddedResource>
   </ItemGroup>
   ```
3. Create another project (Project B) that has a project reference to Project A.
4. Create another project (Project C) that has a project reference to Project B.
5. Build the projects.

### Expected Behavior

The satellite assembly for the culture `yue` is correctly copied into Project B and Project C, and also written into their `.deps.json`.

### Actual Behavior

- The satellite assembly for the culture `yue` is not copied into Project C at all.
- The `.deps.json` in both Project B and Project C do not include the satellite assembly for the culture `yue`.

### Analysis

Suspicious related method: [`ReferenceTable.FindSatellites()`](https://github.com/dotnet/msbuild/blob/4bb8d030c8bb0ffd673c031947251155ecaac390/src/Tasks/AssemblyDependency/ReferenceTable.cs#L950)

https://github.com/dotnet/msbuild/blob/4bb8d030c8bb0ffd673c031947251155ecaac390/src/Tasks/AssemblyDependency/ReferenceTable.cs#L974

This method cannot find satellite assemblies with unusual cultures because it is still checking the satellite directory name against `CultureInfoCache` only.

Related issues and PRs:
- #9954
  (Fixed by #10026)

### Versions & Configurations

MSBuild version: 17.11.2.32701

### PR DESCRIPTION
**Pull request by JanKrivanek**
Fixes #10536
Contributes to https://github.com/dotnet/msbuild/issues/10996

### Context
Custom culture on EmbeddedResource items is supported by `AssignTask` (from https://github.com/dotnet/msbuild/pull/10026), but it's then ignored by RAR task (and CreateManifestResourceName as well) 

### Changes Made
We are no more checking the recoginzed culture against the known list. Putting this behind changewave - as it has pottential to break esoteric scenarios of binaries with .resource.dll existing in subdirs, but actually not being a satelite assemblies

### Testing
Manual testing on a scenario from the bug
Tailored tests added in https://github.com/dotnet/msbuild/pull/11023 (will be uncommented after merging this PR)

### CODE REVIEW

**Comment by maridematte** [comment]
For changes featured in ChangeWaves please update the [changewave doc](https://github.com/dotnet/msbuild/blob/8f6b8ad0ace90c777c66711c907227fcfb6f2efe/documentation/wiki/ChangeWaves.md).

File: src/Tasks/AssemblyDependency/ReferenceTable.cs [hunk 1]
```diff
@@ -971,7 +971,8 @@ private void FindSatellites(
                     // Is there a candidate satellite in that folder?
                     string cultureName = Path.GetFileName(subDirectory);
 
-                    if (CultureInfoCache.IsValidCultureString(cultureName))
+                    // Custom or unknown cultures can be met as well
+                    if (ChangeWaves.AreFeaturesEnabled(ChangeWaves.Wave17_14) || CultureInfoCache.IsValidCultureString(cultureName))
```
Comment (SimaTian) [inline src/Tasks/AssemblyDependency/ReferenceTable.cs:975]: So after the change wave is rotated out, we can discard this if altogether?

File: src/Tasks/AssignCulture.cs [hunk 2]
```diff
@@ -3,6 +3,7 @@
 
 using System;
 using System.Collections.Generic;
+using Microsoft.Build.Collections;
```
Comment (SimaTian) [inline src/Tasks/AssignCulture.cs:6]: I don't see any collection in the file changes. What is this used for please?

File: src/Tasks/AssemblyDependency/ReferenceTable.cs (hunk 1 above)
Comment (JanKrivanek) [reply src/Tasks/AssemblyDependency/ReferenceTable.cs:975]: Exactly.
Whenever removing stale changewaves, it's expected to remove conditionality that becames peremanent (including  removing code that becames unreachable - not a case here though)

File: src/Tasks/AssignCulture.cs (hunk 2 above)
Comment (JanKrivanek) [reply src/Tasks/AssignCulture.cs:6]: The `MSBuildNameIgnoreCaseComparer` added is defined there

File: src/Tasks/CreateCSharpManifestResourceName.cs [hunk 3]
```diff
@@ -101,12 +101,25 @@ internal static string CreateManifestNameImpl(
...
+                info = new Culture.ItemCultureInfo()
+                {
+                    culture = culture,
+                    cultureNeutralFilename =
+                        embeddedFileName.RemoveLastInstanceOf("." + culture, StringComparison.OrdinalIgnoreCase)
+                };
+            }
+            else
+            {
+                info = Culture.GetItemCultureInfo(embeddedFileName, dependentUponFileName, treatAsCultureNeutral);
```
Comment (surayya-MS) [inline src/Tasks/CreateCSharpManifestResourceName.cs:117]: why do we need this line?
```info = Culture.GetItemCultureInfo(embeddedFileName, dependentUponFileName, treatAsCultureNeutral);```

**Review by surayya-MS — APPROVED** [review]
Looks good to me!

File: documentation/wiki/ChangeWaves.md [hunk 4]
```diff
@@ -33,6 +33,7 @@ A wave of features is set to "rotate out" (i.e. become standard functionality) t
 - [Add ParameterName and PropertyName to TaskParameterEventArgs](https://github.com/dotnet/msbuild/pull/10130)
 - [Emit eval props if requested by any sink](https://github.com/dotnet/msbuild/pull/10243)
 - [Load Microsoft.DotNet.MSBuildSdkResolver into default load context (MSBuild.exe only)](https://github.com/dotnet/msbuild/pull/10603)
+- [Support custom culture in RAR](https://github.com/dotnet/msbuild/pull/11000)
```
[the remark this answers is not in the document]
Comment (JanKrivanek) [reply documentation/wiki/ChangeWaves.md:36]: Thanks!
Moved

File: src/Framework/StringUtils.cs [hunk 5]
```diff
@@ -31,4 +31,24 @@ internal static string GenerateRandomString(int length)
...
+    /// <summary>
+    /// Removes last occurence of <paramref name="substring"/> from <paramref name="fromString"/>, if present.
+    /// </summary>
+    /// <param name="fromString">String to be altered.</param>
+    /// <param name="substring">String to be removed.</param>
+    /// <param name="comparison">The comparison to use for finding.</param>
+    /// <returns>The original string (if no occurrences found) or a new string, with last instance of <paramref name="substring"/> removed.</returns>
+    internal static string RemoveLastInstanceOf(this string fromString, string substring, StringComparison comparison = StringComparison.Ordinal)
+    {
+        int lastOccurrenceIndex = fromString.LastIndexOf(substring, StringComparison.Ordinal);
```
[the remark this answers is not in the document]
Comment (JanKrivanek) [reply src/Framework/StringUtils.cs:44]: applied

File: src/Tasks/CreateCSharpManifestResourceName.cs (hunk 3 above)
Comment (JanKrivanek) [reply src/Tasks/CreateCSharpManifestResourceName.cs:117]: It's the original behavior of this method - so needed when changewave is opted out.

Also - if no explicit culture is specified - we still should try to infer it

**Comment by jwoodward10** [comment]
This code seems to be deployed with MSBuild version 17.13.19+0d9f5a35a for .NET Framework. Why isn't the https://learn.microsoft.com/en-us/visualstudio/msbuild/change-waves?view=vs-2022 document up to date with the commits in this PR?

**Pull request merged**